Add support for SeratoMarkers2 GEOB tag (Hotcues/Loops/Flips/etc.)

Holzhaus opened the pull request:
This is a WIP to add support for the SeratoMarkers2 GEOB tag. This tag contains cuepoints, saved loops, "flips" and other information and is used by the Serato DJ Pro software.

It's nowhere near completion, but due to lack of free time I was unable to invest a lot of time into this. Help is very much appreciated.

I'm unsure how to integrate support properly. I'd like Mixxx to read cues from the tag when a track is loaded into a deck (or its cues are edited via the right click dialog), but also write tags back to the file if anything is changed (this can be made optional).

This is not only useful for former Serato users that have a big library and want to migrate cue points to Mixxx, but also for use cases where multiple computers are used: I usually play around with Mixxx on my desktop PC (testing out transitions, settings cuepoints, etc). Before a gig, I then rsync the tracks I want to play over to my laptop. However, all cuepoints are lost in the process (I could copy over the library, search/replace all paths and remove non-copied tracks but this is cumbersome).
Support for the SeratoMarkers2 tag would allow me to just copy over the MP3 file without losing cuepoints.

Launchpad bug: https://bugs.launchpad.net/mixxx/+bug/741613
Writeup: http://homepage.ruhr-uni-bochum.de/jan.holthuis/posts/reversing-seratos-geob-tags
Documentation: https://github.com/Holzhaus/serato-tags/blob/master/docs/serato_markers2.md

Holzhaus committed:
track: Add support for GEOB frames

Holzhaus committed:
track: Add preliminary parsing of "SeratoMarkers 2" tag contents

Holzhaus committed:
track: Add proper SeratoMarkers2 parsing

Be-ing commented:
Thank you for publishing this. Hopefully someone else can pick up where you left off.

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -153,10 +156,20 @@ const QString ID3V2_TYER_FORMAT("yyyy");
     if (tString.isNull()) {
         // null -> null
-        return QString();
+        return QByteArray();
```
This change doesn't look like it was intended?

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -583,6 +617,57 @@ int removeUserTextIdentificationFrames(
     return count;
 }
 
+// Finds the first text frame that with a matching description (case-insensitive).
```
Copy&Paste: "GEOB", neither "text" nor "comments" applies here

uklotzde commented:
I'll provide a fixed and rebased version soon. Waiting for CI builds.

uklotzde committed:
Fix backwards compatibility with older Qt versions

Holzhaus commented:
@uklotzde I added you as a collaborator to my `mixxx` repository to grant you write permissions (in case you don't want to open a new PR).

uklotzde commented:
https://github.com/uklotzde/mixxx/tree/serato-markers2

- Rebased on master
- Code for reading/writing of GEOB tags hidden behind the build flag \__EXTRA_METADATA\__
- Added workarounds for Qt <5.12, i.e. the branch should build on Xenial (Travis CI)
- Minor fixes and cleanup

@Holzhaus I could do a _force push_ into your repo, but maybe you first make a backup of your original branch before replacing it.

Holzhaus commented:
@uklotzde I created a copy of this branch here: https://github.com/Holzhaus/mixxx/tree/serato-markers2-backup

uklotzde commented:
We can merge this PR right now to complete and enable this feature later. Importing and parsing of tags are still disabled.

Any thoughts?

Holzhaus committed:
track/seratomarkers2: Add support for BPMLOCK entries

Holzhaus committed:
track/seratomarkers2: Add support for COLOR entries

Holzhaus committed:
track/seratomarkers2: Add support for LOOP entries

Holzhaus committed:
track/seratomarkers2: Check minimum length of cue entry

Be-ing commented:
What is the point of merging code that is not useful yet?

uklotzde commented:
@Be-ing To keep it up-to-date and allow someone to pick it up? Letting it rot in a PR that piles up merge conflicts would be a pity. This is definitely something that will be needed eventually.

Be-ing commented:
Okay, as long as there are some comments in the code documenting what the current state is and giving some hints on next steps to whoever may pick it up.

Holzhaus committed:
track/seratomarkers2: Add terminating null byte after CUE entry label

Holzhaus committed:
track/seratomarkers2: Add terminating null byte after LOOP entry label

Holzhaus committed:
track/seratomarkers2: Fix LOOP entry magic value check

Holzhaus committed:
track/seratomarkers2: Print warning messages on parsing errors

Holzhaus committed:
test: Add tests for SeratoMarkers2 feature

Holzhaus committed:
test: Add example data and parsing tests for seratomarkers2

Holzhaus committed:
track/seratomarkers2: Add data() method for SeratoMarkers2 instances

Holzhaus committed:
track/seratomarkers2: Fix UTF-8 label length calculation of CUE/LOOP entries

Holzhaus committed:
track/seratomarkers2: Fix indentation of multi-line qWarning() messages

Holzhaus committed:
track/seratomarkers2: Improve parser warnings in SeratoMarkers2 class

Holzhaus committed:
track/seratomarkers2: Detect trailing null byte in entry list

Holzhaus committed:
track/seratomarkers2: Remove last byte of b64 data if padding is necessary

Holzhaus committed:
track/seratomarkers2: Remember "allocated size" to reduce disk fragmentation

Holzhaus committed:
test: Check if SeratoMarkers2 can do a lossless parse/unparse round-trip

Holzhaus committed:
test: Print filename of SeratoMarkers2 dump to simplify debugging

Holzhaus commented:
I added a bunch of tests to make sure that reading/writing the `SeratoMarkers2` binary format works. While working on this I also found some bugs that are now fixed. The data files added by commit ec1f38210b5a18fb3884776e02d0e20e280f73d7 are taken from [my serato-tags repository](https://github.com/Holzhaus/serato-tags/tree/master/data). These are real-world examples of data written by Serato DJ Pro.

I also added support for more entry types. Besides `CUE` (hotcues), this can also parse `BPMLOCK` (determines if the beatgrid is locked), `COLOR` (stores the track's color in the tracklist) and `LOOP` (saved loops) entries now. Essentially everything except "FLIP" entries are parsed now (these are a bit more complicated and also don't map to any functionality present in Mixxx, so I didn't bother to add them). As before, unknown entry types like `FLIP` are not parsed but still stored in the entry list, so they don't get lost when modifying a tag.

So except for parsing `FLIP` entries, the "Serato side" of this is more or less done.

What is still missing is the proper integration into Mixxx:
- Conversion between `CUE` entries and Mixxx' cues
- Integrating cues read from `SeratoMarkers2` into the tracks cue list (and likewise taking Mixxx' cues and writing them in the `SeratoMarkers2` instance)
- Actually writing modified data into the file tags
- A setting to toggle reading/writing this tag on/off in the preferences dialog

If you want me to add `//TODO:` comments that explain this, please tell me where to add them. I haven't found a fitting location.

Be-ing commented:
Great work! Thanks for the summary. I don't think we _need_ it in the code, but having it written here helps.

Be-ing commented:
Have you checked if Serato DJ Intro and Serato DJ Pro have any differences in these tags? I doubt they do, but it would be a good idea to verify that.

Holzhaus commented:
I didn't, but I'll check. Even if that's the case, there are still the `\x01\x01` bytes at the beginning of the tag which are some kind of version number I suspect. If the parser encounters any issues (e.g. if the initial bytes are NOT `\x01\x01`), it'll abort parsing and print a error message on the terminal.
If Mixxx encounters such a case, it should just ignore the tag completely and also disable writing it so that no tag data is lost.

If you have some tracks with Serato DJ Intro tag data at hand, you can test this yourself. Just use eyeD3 to dump the data and put it into the `src/test/seratomarkers2-data` directory and run the tests:

```bash
$ eyeD3 --write-objects /tmp/dump /path/to/file.mp3
$ cp /tmp/Serato\ Markers2.octet-stream src/test/seratomarkers2-data/mydump.octet-stream
$ ./mixxx-test --gtest_filter=SeratoMarkers2Test.ParseMarkers2Data

Holzhaus commented:
The CI fail is unrelated (unmet dependency on OSX).

Holzhaus commented:
@Be-ing I just checked: Serato DJ Pro, Serato DJ Lite and Serato DJ Intro all use this format to store their information.

uklotzde commented:
@Be-ing @Holzhaus I would like to merge this soon, because I have some metadata handling improvements ready that would cause merge conflicts and need to be resolved first.

Be-ing commented:
That's fine with me if @Holzhaus is okay with merging it at this point.

Holzhaus commented:
Yup, this is disabled anyway by default, parse/unparse tests are in place and pass. Feel free to merge.

uklotzde commented:
LGTM.

Tests are in place and the extra code only adds a small compile time but no runtime overhead.

Holzhaus commented:
By the way, this only adds support Serato data stored in for ID3v2 tags. I started to document other file formats here: https://github.com/Holzhaus/serato-tags/blob/master/docs/fileformats.md

uklotzde commented:
CI builds: Currently we don't have any build option in master that needs and enables those extra tags. But #2282 is in place and I will rebase it continually on master.

bungrudi commented:
I really feel that this work deserve much more attention.

To give some context, I just started into DJ-ing and looking into Serato alternative that will allow me to leverage my programming skill (i.e. into programming MIDI script/automation). I'm a programmer in my day job and naturally feel that I should be able to automate many things (... which currently requires kung-fu skill to pull off in Serato), especially that I have a Novation Launchpad as a secondary controller.

I just found and tried Mixxx yesterday and immediately felt that this (the interoperability with Serato database) is exactly the missing feature that this app need, and what prevented me to use it currently.

My controller came with Serato license and I invested a lot of time and money on the software. I have all my tracks tagged on Serato (hotcue, hot loops, keys, etc). I invested in a tool that will analyze and put automatic cue points as Serato data (Mixed in Key, $58).

VirtualDJ is almost perfect, except that its scripting language is super clunky and the license price is steep at $299.
Having VirtualDJ on trial for a month, I can almost say for certainty that one reason for its success is its ability to work with Serato data format; in both direction.

My hunch tells me that Mixxx will took off once it implements the same feature. 

BTW good work @Holzhaus . 

> @Be-ing To keep it up-to-date and allow someone to pick it up? Letting it rot in a PR that piles up merge conflicts would be a pity. This is definitely something that will be needed eventually.

Be-ing commented:
@bungrudi if you want to work on completing this, that would be great! @Holzhaus outlined what is left to do [above](https://github.com/mixxxdj/mixxx/pull/2323#issuecomment-545228335).

bungrudi commented:
@Be-ing I'm a spoilt Java + C# coder.. you don't want me to introduce leaks :D 
Can I somehow code it in a sandboxed, ephemeral environment i.e. in the Javascript engine ?

Be-ing commented:
No, this would have to be done by editing Mixxx's C++ code. I think if you know Java and C# it would not be so difficult to pick up C++. We have [guides](https://mixxx.org/wiki/doku.php/developer_tools#using_ides) on the Mixxx wiki for setting up some IDEs to work on Mixxx. Even if you were an experienced C++ coder, we would review your code before merging it, so don't let making stupid mistakes stop you.

bungrudi commented:
I started with C++, but its been 10 years so please understand that I will be very rusty. Also this will be a a weekend activity for me so if I do pick it up you won't see any commit from me until the next 2 weeks probably.
I'll branch and setup my environment later today and see what I can do.

Be-ing commented:
That's fine; we're all volunteers here too. Feel free to ask on our [Zulip chat](https://mixxx.zulipchat.com/) if you need help getting started. I think we will be busy working on other features anyway until you have some commits to show.

Fortunately C++ has improved a bit since 10 years ago. You don't need to worry about most fancy new language features that are rarely used, but some common tasks are nicer now (like iterating container objects and having an explicit `nullptr` value). You will probably need to refer to the [Qt documentation](https://doc.qt.io/qt-5/classes.html) when working on Mixxx.

uklotzde commented:
@bungrudi Please note that I have some pending metadata updates and fixes queued in my personal branch [dev_library_metadata](https://github.com/uklotzde/mixxx/tree/dev_library_metadata) (frequently rebased) that depend on PR #2365.

Be-ing commented:
How does that impact the work to be done to complete this feature? Should @bungrudi start from #2365? Or start from your dev_library_metadata branch?

uklotzde commented:
Since I am rebasing this branch often I don't recommend to base new work on it now. Just check frequently that no merge conflicts arise.

Be-ing commented:
It looks like #2365 will make a big impact on the remaining tasks for this feature. I think it would make sense for @bungrudi to start working from that or wait for that to be merged. I will prioritize reviewing #2365 to not hold up @bungrudi's work.

bungrudi commented:
Noted that @uklotzde . 
Currently I am basing on master, let me know if i have to rebase to #2365 @Be-ing .

uklotzde commented:
@bungrudi You can check for arising merge conflicts by trying to merge the corresponding branch from time to time. Just revert back to your last commit afterwards.

Holzhaus commented:
@bungrudi If you don't want to deal with merge conflict problems and reverting commits just do this:

```bash
$ git remote add uklotzde https://github.com/uklotzde/mixxx.git  # Only do this once
# Develop you stuff in branch "my-feature-branch". When you want to test if your changes cause merge conflicts, do this:
$ git checkout -b my-feature-branch-mergetest  # Branch off from your feature branch
$ git fetch uklotzde && git merge uklotzde/dev_library_metadata  # Merge changes from remote uklotzde
$ git merge --abort  # Only do this if a merge conflict occurs
$ git checkout my-feature-branch  # Switch back to your feature branch
$ git branch -D my-feature-branch-mergetest  # Delete the mergetest branch
```

Be-ing commented:
@bungrudi if you did not see, #2365 was merged to master.

Holzhaus commented:
@bungrudi Did you already start working on the integration?

Holzhaus commented:
@bungrudi If you're still interested, here's a first quick&dirty hack to read hotcues from metadata: https://github.com/mixxxdj/mixxx/compare/master...Holzhaus:serato-markers2-integration
